onboarding-manager/internal/handlers/controller/reconcilers/instance.go

```diff
@@ -85,9 +86,70 @@ func (ir *InstanceReconciler) Reconcile(ctx context.Context,
 		return directive
 	}
 
+	if directive := ir.handleHostDeauthorized(ctx, instance, request, resourceID); directive != nil {
+		return directive
+	}
+
 	return ir.reconcileInstance(ctx, request, instance)
 }
 
+func checkStatusIdle(instance *computev1.InstanceResource,
+) bool {
+	// Check if all statuses in instance are Idle
+	if instance.GetInstanceStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_IDLE &&
+		instance.GetInstanceStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_UNSPECIFIED {
+		return false
+	}
+	if instance.GetProvisioningStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_IDLE &&
+		instance.GetProvisioningStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_UNSPECIFIED {
+		return false
+	}
+	if instance.GetUpdateStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_IDLE &&
+		instance.GetUpdateStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_UNSPECIFIED {
+		return false
+	}
+	if instance.GetTrustedAttestationStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_IDLE &&
+		instance.GetTrustedAttestationStatusIndicator() != statusv1.StatusIndication_STATUS_INDICATION_UNSPECIFIED {
+		return false
+	}
+	return true
+}
+
+func (ir *InstanceReconciler) handleHostDeauthorized(ctx context.Context, instance *computev1.InstanceResource,
+	request rec_v2.Request[ReconcilerID], resourceID string,
+) rec_v2.Directive[ReconcilerID] {
+	if instance.GetHost().GetCurrentState() == computev1.HostState_HOST_STATE_UNTRUSTED ||
+		instance.GetHost().GetDesiredState() == computev1.HostState_HOST_STATE_UNTRUSTED {
+		checkCleanupRun := false
+		// Check that all statuses and indicators have been updated
+		if !checkStatusIdle(instance) {
+			// Update instance statuses to idle
+			util.PopulateInstanceIdleStatus(instance)
+			checkCleanupRun = true
+		}
+		// If the host associated with the instance is deauthorized, check if the provisioning
+		// workflow has been removed and clean it up if not
+		deviceInfo, err := convertInstanceToDeviceInfo(instance)
+		if err == nil && onboarding.CheckWorkflowExist(ctx, deviceInfo, instance) {
+			// Delete workflow and set provisioning status
+			if err := ir.cleanupProvisioningResources(ctx, instance); err != nil {
+				// If error received, don't retry as resources will be deleted when instance is deleted
+				return request.Ack()
+			}
+			checkCleanupRun = true
+		}
+		if checkCleanupRun {
+			// Run reconciliation on the instance to complete
+			zlogInst.Info().Msgf("Host associated with Instance (%s) has been deauthorized. "+
+				"Forcing reconciliation to update Instance status.", resourceID)
+			return ir.reconcileInstance(ctx, request, instance)
```

Comment:
use requeue or resubmti directive?

Reply:
Since we have removed the deauth of the instance from this, we can return `request.Ack()` instead after updating the statuses and removing the workflows